src/cryptsetup/cryptsetup-token-systemd-tpm2/cryptsetup-token-systemd-tpm2.c

```diff
@@ -0,0 +1,223 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <libcryptsetup.h>
+
+#include "cryptsetup-token.h"
+#include "hexdecoct.h"
+#include "memory-util.h"
+#include "tpm2-util.h"
+#include "luks2-tpm2.h"
+
+#define TOKEN_NAME "systemd-tpm2"
+#define TOKEN_VERSION_MAJOR "1"
+#define TOKEN_VERSION_MINOR "0"
+
+#define crypt_log_dbg(cd, x...) crypt_logf(cd, CRYPT_LOG_DEBUG, x)
+#define crypt_log_err(cd, x...) crypt_logf(cd, CRYPT_LOG_ERROR, x)
+#define crypt_log_std(cd, x...) crypt_logf(cd, CRYPT_LOG_NORMAL, x)
+
+/* for libcryptsetup debug purpose */
+_public_ const char *cryptsetup_token_version(void) {
+        return TOKEN_VERSION_MAJOR "." TOKEN_VERSION_MINOR;
+}
+
+static int log_debug_open_error(struct crypt_device *cd, int r) {
+        if (r == -EAGAIN)
+                crypt_log_dbg(cd, "TPM2 device not found.");
+        else if (r == -ENXIO)
+                crypt_log_dbg(cd, "No matching TPM2 token data found.");
+        else if (r == -ENOMEM)
+                crypt_log_dbg(cd, "Not Enough memory.");
+        else if (r == -EINVAL)
+                crypt_log_dbg(cd, "Internal error unlocking device using system-tmp2 token.");
```

Comment:
Shouldn't there be a final `else` branch printing a generic log message for all other errors here?